gem/optimise.py

```diff
@@ -413,3 +480,565 @@ def aggressive_unroll(expression):
+
+
+@_count_flop.register(MathFunction)
+@_count_flop.register(LogicalNot)
+@_count_flop.register(LogicalAnd)
+@_count_flop.register(LogicalOr)
+def _count_flop_func(node, self):
+    return self(node.children[0]) + 1
+
+
+@_count_flop.register(Conditional)  # this is not quite right
+@_count_flop.register(Power)
+@_count_flop.register(Comparison)
+@_count_flop.register(Sum)
+@_count_flop.register(Product)
+@_count_flop.register(Division)
+def _count_flop_common(node, self):
+    flop = numpy.product([i.extent for i in node.free_indices])
+    # Hoisting the factors
+    for child in node.children:
+        flop += self(child)
+    return flop
+
+
+@_count_flop.register(Constant)
+@_count_flop.register(Terminal)
+@_count_flop.register(Indexed)
+@_count_flop.register(FlexiblyIndexed)
+def _count_flop_const(node, self):
+    return 0
+
+
+def count_flop(node):
+    mapper = Memoizer_cf(_count_flop)
+    return mapper(node)
+
+
+@singledispatch
+def _expand_all_product(node, self, index):
+    # ---------------------------------------
+    # Expand all products recursively if free index of node include index
+    # from :param:`index`
+    # e.g (a+(b+c)d)e = ae + bde + cde
+
+    raise AssertionError("cannot handle type %s" % type(node))
+
+
+_expand_all_product.register(Node)(reuse_if_untouched_arg)
+
+
+@_expand_all_product.register(Product)
+def _expand_all_product_common(node, self, index):
+    a = self(node.children[0], index)
+    b = self(node.children[1], index)
+    if isinstance(b, Sum) and any([i in b.free_indices for i in index]):
+        return Sum(self(Product(a, b.children[0]), index),
+                   self(Product(a, b.children[1]), index))
+    elif isinstance(a, Sum) and any([i in a.free_indices for i in index]):
+        return Sum(self(Product(a.children[0], b), index),
+                   self(Product(a.children[1], b), index))
+    else:
+        return node
+
+
+def expand_all_product(node, index):
+    mapper = MemoizerArg(_expand_all_product)
+    return mapper(node, index)
+
+
+def collect_terms(node, node_type):
+    """Recursively collect all children into a list from :param:`node`
+    and its children of class :param:`node_type`.
+
+    :param node: root of expression
+    :param node_type: class of node (e.g. Sum or Product)
+    :return: list of all terms
+    """
+    from collections import deque
+    terms = []  # collected terms
+    queue = deque([node])  # queue of children nodes to process
+    while queue:
+        child = queue.popleft()
+        if isinstance(child, node_type):
+            queue.extendleft(reversed(child.children))
+        else:
+            terms.append(child)
+    return tuple(terms)
+
+
+def flatten_sum(node, argument_indices):
+    """
+    factorise :param:`node` into sum of products, group factors of each product
+    based on its dependency on index:
+    1) nothing (key = 0)
+    2) i (key = 1)
+    3) (i)j (key = j)
+    4) (i)k (key = k)
+    ...
+    5) (i)jk (key = 2)
```

Comment:
This was fine back in the age of FIAT, when we had one quadrature index (i) and at most two argument/linear indices (j and k). Since the dawn of FInAT, however, the situation is much different. Currently you can have up to 3 quadrature indices, and up to 6, 8, 10 or in theory even more argument indices.